### PR TITLE
LIMS-1074: Allow accessing zipped files

### PR DESCRIPTION
**Pull request by ndg63276**
**JIRA ticket**: [LIMS-1074](https://jira.diamond.ac.uk/browse/LIMS-1074)

**Summary**:

To save on disk space, many files created by auto processing pipelines have been zipped (*.gz), which breaks the links to them in Synchweb, as well as a number of image files and map / model viewers. These should be fixed server-side.

**Changes**:
- Use gzopen and gzread to temporarily unzip the processing files before serving them to the front end
- Add already zipped files to archive.zip file served by "Archive" button (they download as unzipped files zipped together)
- Fix the map / model viewer shell script to unzip the relevant files first
- Fix 2 silly spelling mistakes

**To test**:
- View a pre-2022 data collection with autoprocessing, eg /dc/visit/mx23694-64/id/6669874
- For fast_dp
  - Check the 'Archive' button downloads a full zip file
  - Check the 'Plots' button produces graphs
  - Check the 'Radiation Damage' button produces a graph
  - View the 'Logs & Files' modal, and try viewing files of type log and graph and downloading files of type log, graph and result
  - Repeat tests for different autoprocessing pipelines
- For downstream processing
  - Check the 'Archive' and 'Logs & Files' buttons as above
  - Check the 'Map / Model Viewer' button loads a model
  - Check thumbnails of models load for Autosharp, Autobuild and Crank2
- Check no error messages appear on display or in dev console

### CODE REVIEW

File: api/src/Downstream/Type/FastEp.php [hunk 1]
```diff
@@ -17,82 +17,98 @@ function results() {
...
-                            $parts[7],
-                            $parts[8],
-                        ));
-                    }
-                }
+            } elseif (file_exists($pdb['FILE'].'.gz')) {
+                $zd = gzopen($pdb['FILE'].'.gz', 'r');
+                $pdb = gzread($zd, 10000000);
+                gzclose($zd);
+            } else {
```
Comment (gfrn) [inline api/src/Downstream/Type/FastEp.php:24]: This is limited to 10^7 bytes of uncompressed data, which might not be enough for some files. An alternative data stream/loading strategy should be considered.

**Review by John-Holt-Tessella — COMMENTED** [review]
Not sure how urgent these thing are. I will go test ome stuff :-)

**Review by John-Holt-Tessella — CHANGES_REQUESTED** [review]
Files in archives (once downloaded) appear to be files but are actually the zipped version of those files.

File: api/src/Page/Download.php [hunk 2]
```diff
@@ -474,7 +476,11 @@ function _streamZipFile($files, $zipName)
             $zip = new ZipStream($zipName . '.zip', $options);
             foreach ($files as $file) {
                 $filename = $file['FILEPATH'] . '/' . $file['FILENAME'];
-                $zip->addFileFromPath(basename($filename), $filename);
+                if (file_exists($filename)) {
+                    $zip->addFileFromPath(basename($filename), $filename);
+                } elseif (file_exists($filename.'.gz')) {
+                    $zip->addFileFromPath(basename($filename), $filename.'.gz');
```
Comment (John-Holt-Tessella) [inline api/src/Page/Download.php:482]:
```suggestion
                    $zip->addFileFromPath(basename($filename).'.gz', $filename.'.gz');
```

File: api/scripts/mtz2map.sh [hunk 3]
```diff
@@ -15,6 +15,24 @@ export CLIBD=$CCP4_MASTER/lib/data
 export CCP4_SCR=/tmp
 export root=$CCP4_MASTER/bin
 
+if [ -f $1 ]; then
+	mtz=$1
+else
+	if [ -f $1.gz ]; then
+		mtz=/tmp/$2_$3.mtz
+		gunzip -c $1.gz > $mtz
```
[the remark this answers is not in the document]
Comment (ndg63276) [reply api/scripts/mtz2map.sh:23]: Now deleted on line 106

File: api/scripts/mtz2map.sh [hunk 4]
```diff
@@ -15,6 +15,24 @@ export CLIBD=$CCP4_MASTER/lib/data
...
+	fi
+fi
```
[the remark this answers is not in the document]
Comment (ndg63276) [reply api/scripts/mtz2map.sh:28]: Now exits if no mtz

File: api/scripts/mtz2map.sh [hunk 5]
```diff
@@ -15,6 +15,24 @@ export CLIBD=$CCP4_MASTER/lib/data
...
+
+if [ -f $4 ]; then
+	pdb=$4
+else
+	if [ -f $4.gz ]; then
+		pdb=/tmp/$2_$3.pdb
```
[the remark this answers is not in the document]
Comment (ndg63276) [reply api/scripts/mtz2map.sh:36]: Now deleted on line 90

File: api/scripts/mtz2map.sh [hunk 6]
```diff
@@ -15,6 +15,24 @@ export CLIBD=$CCP4_MASTER/lib/data
...
+		gunzip -c $4.gz > $pdb
+	fi
+fi
```
[the remark this answers is not in the document]
Comment (ndg63276) [reply api/scripts/mtz2map.sh:42]: Now exits if it doesnt exist (within if clause for dimple/mrbump)

File: api/src/Downstream/Type/FastEp.php [hunk 7]
```diff
@@ -17,82 +17,94 @@ function results() {
...
-                            $mat
-                        )
-                    ) {
-                        $dat['FOM'] = floatval($mat[1]);
-                        $dat['CC'] = floatval($mat[2]);
-                    }
+            } elseif (file_exists($lst['FILE'].'.gz')) {
+                $lst = gzdecode(file_get_contents($lst['FILE'].'.gz'));
+            } else {
+                $this->_error('Could not find sad.lst file');
```
[the remark this answers is not in the document]
Comment (ndg63276) [reply api/src/Downstream/Type/FastEp.php]: Now should be the same behaviour, ie skip this block if file or file.gz doesnt exist

File: api/src/Page/Download.php [hunk 8]
```diff
@@ -254,31 +257,30 @@ function _get_file($id, $file)
     {
         // We don't want to allow unlimited file sizes
         ini_set('memory_limit', '512M');
-        $filesystem = new Filesystem();
```
[the remark this answers is not in the document]
Comment (ndg63276) [reply api/src/Page/Download.php]: Have put the symfony version back in

File: api/src/Page/Download.php (hunk 2 above)
Comment (ndg63276) [reply api/src/Page/Download.php]: Rats, I wondered why we didn't get a zip of zips. Have committed the suggestion.

**Review by John-Holt-Tessella — APPROVED** [review]

**Pull request merged**